Adding a new filter plugin that can encrypt and decrypt values using AWS KMS service

achinthagunasekara opened the pull request:
## SUMMARY

Adding a new filter plugin that can encrypt and decrypt values using AWS KMS service.
This filter plugin can be used to encrypt values using the AWS KMS service and store them within your repository, then decrypt then when you want to use them.

## Example

#### Decrypting

In `my_role\defaults\main.yml`
```
encrypted_value: >-
  {{ abcd123abcd13 abcd123abcd13abcd12abcd13
  abcd123abcd13abcd123abcd13abcd123ab1d1312
  abcd123abcd13abcd123abcd13abcd123ssdsas12
  sadsdsasd | aws_kms_decrypt(KEY_ARN) }}
```

In `my_role\tasks\main.yml`

```
- debug:
    msg: "Decrypted value is {{ encrypted_value }}"
```

Same can be done for encrypting values.

#### Encrypting

In `my_role\tasks\main.yml`

```
- debug:
    msg: "Encrypted value is {{ 'abc123abc123abc123' | aws_kms_encrypt(KEY_ARN) }}"
```

## Arguments

You can call `aws_kms_encrypt` and `aws_kms_encrypt` from this plugin with following variables.

* Mandatory - ciphertext (str): Encrypted item to decrypt.
* Mandatory - key_arn (str): AWS ARN to the KMS key.
* Optional - region_name (str): AWS region to use.
* Optional - profile_name (str): AWS credential profile to use.
* Optional - role_to_assume (str): AWS IAM role to use to get credentials.
* Optional - aws_access_key_id (str): AWS access key to use.
* Optional - aws_secret_access_key (str): AWS secret key to use.

Getting AWS Credentias will use following order.

`profile_name` > `role_to_assume` > `aws_access_key_id` and `aws_secret_access_key` > `instance IAM profile` or `default profile`

## ADDITIONAL INFORMATION

* https://aws.amazon.com/documentation/kms/
* https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/introduction.html

achinthagunasekara committed:
Adding a new plugin for aws_kms to encrypt/decrypt items.

juliedavila commented on lib/ansible/plugins/filter/aws_kms.py:
```diff
@@ -0,0 +1,92 @@
+"""
+(c) 2018, Archie Gunasekara <contact@achinthagunasekara.com>.
+Module to handle encrypting and decrypting of items with AWS KMS.
```
GPLv3 header needs to be added

juliedavila replied:
`# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)`

achinthagunasekara replied:
GPLv3 header added

juliedavila commented on lib/ansible/plugins/filter/aws_kms.py:
```diff
@@ -0,0 +1,92 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
```
Needs a metadata block

```
ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'community'
}
```

achinthagunasekara replied:
Metadata block added

juliedavila commented on lib/ansible/plugins/filter/aws_kms.py:
```diff
@@ -0,0 +1,92 @@
+        )
+        return cycled_plaintext.rstrip()
+    except aws_encryption_sdk.exceptions.AWSEncryptionSDKClientError as kms_ex:
+        raise AnsibleFilterError("Unable to encrypt value using KMS - %s" % kms_ex)
```
I assume you meant to say `Unable to decrypt value using KMS`

achinthagunasekara replied:
Message corrected

juliedavila reviewed (changes requested):
Minor changes needed, as a separate note we probably should create botocore connection helpers in module_utils

achinthagunasekara committed:
Addressing comments added by defionscode.

achinthagunasekara reviewed (commented):
@defionscode ready_for_review

achinthagunasekara committed:
Addressing comments by defionscode

defionscode commented:
just address the ansibot comment about using @ to mention me in the commit message

ryansb commented on lib/ansible/plugins/filter/aws_kms.py:
```diff
@@ -0,0 +1,98 @@
+    """
+    dependency_check()
+    try:
+        session = botocore.session.get_session()
```
Connection to AWS cannot rely on the environment/default profile. Lookup plugins involving AWS have to accept arguments for what region, key, and secret to use like this: https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/lookup/aws_account_attribute.pyhttps://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/lookup/aws_account_attribute.py

achinthagunasekara replied:
@ryansb I have added the ability to pass in a `profile`, `role` or `credentials` to this filter plugin. It's ready_for_review again :)

achinthagunasekara committed:
Adding the ability to pass in a AWS profile/role/credentials as pointed
out by ryansb

achinthagunasekara committed:
Ability to pass in AWS region name and code clean up.

achinthagunasekara committed:
Addressing comments by ansibot.

achinthagunasekara committed:
Adding the GNU header
Updating the function name to be more meaningful
Variable name change
Cleaning up the flow of creating a session
Change the filter to throws an AnsibleFilterError instead of AnsibleError for any exceptions
Fixing invalid dictionary key

achinthagunasekara committed:
Adding better validation to arguments passed in.

achinthagunasekara committed:
Fixing typos

achinthagunasekara commented:
@ryansb could you please review my changes again? :)

ready_for_review

achinthagunasekara commented:
@ryansb could you please review my changes again? :)

ready_for_review

achinthagunasekara commented:
@ryansb could you please review my changes again? :)

ready_for_review

s-hertel commented on lib/ansible/plugins/filter/aws_kms.py:
```diff
@@ -0,0 +1,155 @@
+            key_provider=get_key_provider(key_arn=key_arn, **kwargs)
+        )
+        return base64.b64encode(ciphertext)
+    except AWSEncryptionSDKClientError as kms_ex:
```
When the dependency is not installed the error message isn't the one given on line 88 https://github.com/ansible/ansible/pull/42346/files#diff-8117f6964e26dcf20e852f21b50e2a70R88, but is a NameError since `AWSEncryptionSDKClientError` is undefined. The filter should fail for missing dependencies before using them.

achinthagunasekara replied:
@s-hertel @defionscode 

I've made these changes. Could you please have a look?

s-hertel commented on lib/ansible/plugins/filter/aws_kms.py:
```diff
@@ -0,0 +1,155 @@
+        boto3.Session: Returns a boto3.Session object.
+    """
+    valid_keys = [
+        'profile_name',
```
I'm a little worried about the discrepancy between the keys here and what AWS modules and plugins accept. This should probably be documented as well, so the user doesn't have to read the code to figure out what is expected of them.

juliedavila replied:
Perhaps just keep the keys in sync with other stuff, no?

achinthagunasekara replied:
@s-hertel  @defionscode 

I've made these changes. Could you please have a look?

achinthagunasekara committed:
Restrictions on what arguments can be passed on. Better documntation.

achinthagunasekara commented:
@defionscode, @s-hertel, @ryansb 
Are you guys able to provide any feedback on my last changes?
Thanks,

ready_for_review

defionscode commented:
Why did you close this?

achinthagunasekara commented:
@defionscode I raised this again as https://github.com/ansible/ansible/pull/44014
